Use with_unpark_event() to fix some possible deadlocks in Shared.

dwrensha opened the pull request:
Fixes #330 and adds two tests that fail on the current master branch.

alexcrichton commented on src/future/shared.rs:
```diff
@@ -51,85 +70,77 @@ impl<F> Future for Shared<F>
             }
-            State::Done(Ok(ref e)) => Ok(SharedItem { item: e.clone() }.into()),
-            State::Done(Err(ref e)) => Err(SharedError { error: e.clone() }.into()),
+            None => unreachable!(),
```
I think this panic can get triggered, but we can just ignore the case, right?

I'm thinking of a case where:

* One task wins the race to do the poll (turns off `original_future_needs_poll`)
* The future is unparked, setting `original_future_needs_poll` to true
* Another task wins the race to poll (again turning off `original_future_needs_poll`)
* The first task finishes the future, setting it to none
* The second task then panics when it sees the none

Or maybe I missed something?

alexcrichton replied:
If this is true then I think that the reading of `state` may need to be updated as well

dwrensha replied:
Hm. Yeah, I agree that the panic can be triggered. Worse, if `original_future()` unparks the current task and then causes a `Shared` clone to recursively call `poll()` on the same thread, then that thread might attempt to recursively grab the lock for `original_future`.

alexcrichton replied:
Hm good point yeah. Maybe this could get by with a `try_lock` and if it's missed it just falls through to storing a parked task?

dwrensha replied:
No, that would leave `original_future_needs_poll` in the wrong state.

I think it works to probe the `original_future` mutex before doing the swap on `original_future_needs_poll`, to check whether we are in the recursive case.

dwrensha replied:
I've pushed a commit that should fix this issue and includes a test that covers it.

dwrensha replied:
Hm. On second thought, the new branch that I added does not only apply to the recursive case. If a `poll()` on a separate thread fails to grab the `original_future` mutex, it could now get into a busy wait loop of repeatedly calling `task::park().park()`.

dwrensha replied:
I've updated the code comments in that branch. It does seem unfortunate that the `task::park().unpark()` might cause waiting threads to consume a lot of cpu, but I haven't been able to think of a better solution. Fortunately, the problematic case seems like it would be rare: `original_future.poll()` would need to both call `task::park().unpark()` *and* take a lot of time before returning.

alexcrichton replied:
I think with the recent changes this case of panic can still be hit, but double check me:

* A task gets all the way to `drop(original_future)` below, unlocking the lock. The future is completed and dropped
* Another task comes in an polls, sees the `Waiting` state, acquires the original future lock, sees that a poll is needed (some stray unpark set the flag perhaps)
* The second task comes down to this branch

Does that sound right?

dwrensha replied:
Yes, I agree. I think https://github.com/alexcrichton/futures-rs/pull/373/commits/30d90d7bd497c30a5e6b1a4daf8c0b593a37aab2 fixes this, and then https://github.com/alexcrichton/futures-rs/pull/373/commits/92c89415d614eda6725ab57631df3209cfbe2869 sidesteps the entire issue.

dwrensha commented:
I've updated `Unparker` so that it wraps the `original_future_needs_poll` flag and the `tasks` hashmap in a single mutex. This makes it possible to avoid calling `task::park().unpark()` in the common cases. Calling `task::park().unpark()` is still necessary in the case where `original_future.poll()` wakes up its own task.

I *think* this addresses the concerns brought up so far, but... it's subtle. @alexcrichton, can you find any other potential issues?

alexcrichton commented:
@dwrensha could you elaborate on the motivation for multiple locks here? I'd imagine that keeping just one mutex would simplify this perhaps, no?

alexcrichton commented:
(I'm not *too* worried about the poll-during-unpark case btw, just wanted to weed out other issues before scrutinizing that too much)

dwrensha commented:
> could you elaborate on the motivation for multiple locks here?

Good question! The idea was to avoid locking `state` while `original_future.poll()` is being called. However, I'm not sure why it didn't occur to me that we don't actually need to hold *any* lock during `original_future.poll()`. I've pushed a commit that pushes `original_future` back into `State`, avoiding the need for a separate `original_future` mutex.

alexcrichton commented:
Looks good to me! Threading ownership of the future looks like an elegant solution to solve a number of problems.

So turning now to the performance characteristics, I've got two points of hesitation:

1. The thundering stampede problem for *any* unpark. Unfortunately I think this is unavoidable (e.g. the whole impetus for this PR) at this point, and you can always optimize this by using something like a `oneshot` in the middle which is guaranteed to have zero spurious unparks.
2. After the thundering stampede, possible spinning for awhile by all threads.

On the latter point I think it has specifically do do with the `park().unpark()` business. On an unpark all tasks wake up but if they all hit the lock at the same time then one will start polling and N-1 will enter "busy loops". Let's assume that the future won't be ready for some time, so then N-2 will be in a busy loop, etc. (if that makes sense). 

Well hm now after writing that I realize it's only the case when a bunch of unparks are coming in in which case you sorta have to check for it anyway. So maybe it's not that big of a problem!

I wonder, though, perhaps this could be handled by the original poller? Instead of putting threads into a park/unpark loop, perhaps the owner of the future could atomically lock and ensure that polling isn't required? That way the owner would `loop` around polling the future until the "needs poll" flag is unset.

A scheme like that I think would help make a bit more progress (no threads in a "park loop") although it runs the risk of starving the owner of the future by looping for awhile. That being said, that's a pervasive problem, so I don't think we'd need to solve it specifically here.

dwrensha commented:
> Instead of putting threads into a park/unpark loop, perhaps the owner of the future could atomically lock and ensure that polling isn't required? That way the owner would `loop` around polling the future until the "needs poll" flag is unset.

Do you mean wrap the body of `Shared::poll()` in a `loop {}` that continues until `original_future_needs_poll` is `false`? That sounds wrong to me. If `task::pask().unpark()` gets called during `original_future.poll()`, presumably that means that `original_future` wants to yield control to other tasks running on the thread, but if we add a loop we'll just immediately call `original_future.poll()` again.

I've pushed a commit with a nicer solution: we keep a `Vec` of tasks that will be unconditionally unparked when the state transitions away from `State::Polling`. I had in fact considered this approach previously, but when there were separate locks for `state` and `original_future` it was prone to race conditions. Now that `original_future` is part of `state`, I think it works.

alexcrichton commented:
Just to write down my thoughts, the race condition we're considering here is:

* The future is actively being polled
* An unpark happens, setting the "needs a poll" flag
* Another task concurrently comes in, sees the flag, and doesn't see the future

So in general unparks during a poll are totally normal and happen all the time, they just schedule the task to go back and re-poll the future again. The race here though is we have to be sure to not lose that poll request because it could result in the future getting stuck.

With that in mind, I think the case to worry about is when the future is being polled, an unpark happens, all other tasks miss the future itself, then the task polling the future unlocks the future but then forgets about its instance of `Shared` (somehow). We need to make sure that the request to keep polling the future is communicated to other blocked tasks (if any).

Having written that out, the `Vec` feels right here. It's a list of otherwise blocked tasks. When the original future leaves the block it wakes them all up. I believe the task working on the future may *not* have received an unpark from the future's current unpark as it may not have been blocked yet maybe?

In any case, the `Vec` sounds good to me, thanks!
